---
title: "Intro to Epigenomics COMPLETED"
author: Wade Boohar
date: 11/03/24
updated: 11/07/24
---

```{r setup}
 knitr::opts_knit$set(root.dir = normalizePath("~/490_cluster/analysis_data"))
```

```{r}
if (!require("sesameData", quietly = TRUE))
BiocManager::install("sesameData")

if (!require("sesame", quietly = TRUE))
BiocManager::install("sesame")

if (!require("limma", quietly = TRUE))
BiocManager::install("limma")
```

Load in all necessary packages
```{r}
library(TCGAbiolinks)
library(sesame)
library(sesameData)
library(limma)
```

```{r}
query <- GDCquery(project = "TCGA-BRCA",
                  data.category = "DNA Methylation",
                  data.type = "Methylation Beta Value",
                  platform = "Illumina Human Methylation 450",
                  data.format = "Dataframe")
GDCdownload(query)
methylation450 <- GDCprepare(query)
```

```{r}
methylation_clinical <- as.data.frame(methylation450@colData)
betas <- as.data.frame(methylation450@assays@data@listData)
cpg_sites <- as.data.frame(methylation450@rowRanges@elementMetadata)

column_mask <- ifelse(colnames(methylation_clinical) %in% c('treatments', 'primary_site', 'disease_type'), F, T)
methylation_clinical <- methylation_clinical[,column_mask]
write.csv(methylation_clinical, '~/490_cluster/analysis_data/brca_methylation_clinical.csv')


site_mask <- !grepl('-', cpg_sites$gene) & !grepl(';', cpg_sites$gene) & !is.na(cpg_sites$gene) & complete.cases(betas)
betas <- betas[site_mask,]
cpg_sites <- cpg_sites[site_mask,]

write.csv(betas, '~/490_cluster/analysis_data/brca_methylation_betas.csv')
write.csv(cpg_sites, '~/490_cluster/analysis_data/brca_cpg_sites.csv')
```

```{r}
methylation_clinical <- read.csv('~/490_cluster/analysis_data/brca_methylation_clinical.csv', row.names=1)
betas <- read.csv('~/490_cluster/analysis_data/brca_methylation_betas.csv', row.names=1)
cpg_sites <- read.csv('~/490_cluster/analysis_data/brca_cpg_sites.csv', row.names=1)
```

```{r}
normal_tissue_mask <- methylation_clinical$definition == "Solid Tissue Normal"

methylation_clinical$tumor <- !normal_tissue_mask

mval <- t(apply(betas, 1, function(x) log2(x/(1-x))))

design <- model.matrix(~ tumor, data = methylation_clinical)
fit <- lmFit(mval, design)
fit2 <- eBayes(fit)
```

```{r}

dat <- data.frame(foldchange = fit[["coefficients"]][,2], logPvalue =  -log10(p.adjust(fit2[["p.value"]][,2],method='BY')), geneName = cpg_sites$gene)
dat$threshold <- as.factor(abs(dat$foldchange) < 1)

#Visualization
cols <- c("TRUE" = "grey", "FALSE" = "blue")
ggplot(data=dat, aes(x=foldchange, y = logPvalue, color=threshold)) +
  geom_point(alpha=.2, size=0.6) +
  scale_colour_manual(values = cols) +
  geom_vline(xintercept = 1, colour="#990000", linetype="dashed") + 
  geom_vline(xintercept = - 1, colour="#990000", linetype="dashed") +
  geom_hline(yintercept = 2, colour = "#990000", linetype="dashed") +
  theme(legend.position="none") +
  xlab("Fold Change") +
  ylab("-log10 p value") +
  theme_bw() +
  theme(legend.position = "none")
```

...INSERT DESeq2 Stuff here to generate 'results'...
""""

Load in all necessary packages
```{r}
library(TCGAbiolinks)
library(DESeq2)
library(EnhancedVolcano)
```

*1*
Read in the rna_clinical, rna_genes, and rna_counts dataframes which you made in the "SummarizedExperiment Guided Tutorial" R Notebook

```{r}
rna_clinical <- read.csv("~/490_cluster/analysis_data/BRCA_rna_clinical.csv")
rna_genes <- read.csv("~/490_cluster/analysis_data/BRCA_rna_genes.csv")
rna_counts <- read.csv("~/490_cluster/analysis_data/BRCA_rna_counts.csv", row.names=1)
```


*2*
In this assignment, you will run differential expression analysis comparing patient samples by whether the sample is from a tumor or normal tissue (this is the definition column in rna_clinical). You will need to choose a variable to control for covariance of: age and/or PAM50 subtype (paper_BRCA_Subtype_PAM50). 

Manipulate those columns so that they are ready for differential expression analysis (hint: what kind of variables are they? what data type are they by default? do you need to handle unknown values?) Filter out genes with a total expression across all patients less than 1000.

```{r}
rna_clinical[,"Tissue Type"] <- ifelse(rna_clinical$definition == 'Primary solid Tumor', 'Tumor', ifelse(rna_clinical$definition == 'Solid Tissue Normal', 'Normal', NA))
rna_clinical[, "Age"] <- rna_clinical[, "age_at_index"]
rna_clinical[,"Tissue_Type"] <- factor(rna_clinical[,"Tissue Type"])


rna_clinical[,"PAM50_Subtype"] <- factor(rna_clinical$paper_BRCA_Subtype_PAM50)

for(i in 1:dim(rna_clinical)[1]){
  if(rna_clinical[i, 'Tissue_Type'] == 'Normal' & !is.na(rna_clinical[i, 'Tissue_Type'])) {
    for(j in 1:dim(rna_clinical)[1]){
      if(rna_clinical[i, 'patient'] == rna_clinical[j, 'patient'] & !is.na(rna_clinical[j, 'PAM50_Subtype'])) {
        rna_clinical[i, 'PAM50_Subtype'] = rna_clinical[j, 'PAM50_Subtype']
      }
    }
  }
}


NA_mask <- is.na(rna_clinical[, "Tissue_Type"])
NA_mask <- NA_mask | is.na(rna_clinical[, "Age"])
NA_mask <- NA_mask | is.na(rna_clinical[, "PAM50_Subtype"])

rna_clinical_masked <- rna_clinical[!NA_mask,]
rna_counts_masked <- rna_counts[,!NA_mask]

counts_mask <- rowSums(rna_counts_masked) >= 1000
rna_counts_masked <- rna_counts_masked[counts_mask,]
rna_genes_masked <- rna_genes[counts_mask, ]

```


*3*
Perform the differential expression analysis, All you need to do is fill in the appropriate # terms
```{r}
dds <- DESeqDataSetFromMatrix(countData = rna_counts_masked,
                              colData = rna_clinical_masked,
                              design= ~Age+PAM50_Subtype+ Tissue_Type)

dds_obj <- DESeq(dds)

resultsNames(dds_obj)

results <- results(dds_obj, format = "DataFrame", contrast = c("Tissue_Type", "Tumor" , "Normal"))

results <- data.frame(results)
```

Prepare results dataframe for EnhancedVolcano plotting. Add two columns, "-log10(padj)" and "gene_name". Fill in these columns appropriately.
```{r}
results[, "-log10(padj)"] <- -log10(results$padj)
results[, "gene_name"] <- rna_genes_masked$gene_name
```

""""

```{r}
downregulated <- results[(results$log2FoldChange < -3), 'gene_name']
hypermethylated <- dat[dat$foldchange > 2, 'geneName']
interest_genes <- intersect(downregulated, hypermethylated)
```


```{r}
GENE<-"SCTR"

gene_counts_mask <- rna_genes$gene_name == GENE
gene_betas_mask <- cpg_sites$gene == GENE

rna_clinical_tumor <- rna_clinical$definition == "Primary solid Tumor"
methylation_clinical_tumor <- methylation_clinical$definition == "Primary solid Tumor"

rna_clinical_normal <- rna_clinical$definition == "Solid Tissue Normal"
methylation_clinical_normal <- methylation_clinical$definition == "Solid Tissue Normal"

rna_tumor <- as.numeric(rna_counts[gene_counts_mask, rna_clinical_tumor])
methylation_tumor <- (betas[gene_betas_mask, methylation_clinical_tumor])

rna_normal <- as.numeric(rna_counts[gene_counts_mask, rna_clinical_normal])
methylation_normal <- (betas[gene_betas_mask, methylation_clinical_normal])
```

```{r}
boxplot(rna_normal, rna_tumor, xlab='Group', ylab='Counts', names=c('Normal', 'Tumor'))
```


```{r}
plot(1:dim(methylation_normal)[1], apply(methylation_normal, 1, mean), col='blue', type='h', lwd=3, ylab='Average Beta', xlab='CpG sites')
lines(1.3:(dim(methylation_normal)[1]+0.3), apply(methylation_tumor, 1, mean), col='red', type='h', lwd=3)
```

